common/http/src/batch.rs

```diff
@@ -0,0 +1,630 @@
+        self.stream.as_mut().poll_next(cx)
+    }
+}
+
+impl<T: ?Sized> TimedRequestBatcherStreamExt for T where T: Stream {}
+
+pub type BufferSource = Pin<
+    Box<
+        dyn Stream<Item = Result<IngestBodySerializer, IngestLineSerializeError>>
+            + std::marker::Send
+            + std::marker::Sync,
+    >,
+>;
+
+#[derive(Debug, Error)]
+pub enum TimedRequestBatcherError {
+    #[error("{0}")]
+    BufferError(#[from] IngestLineSerializeError),
+    #[error("No buffers available from buffer source")]
+    BufferStreamError,
+}
+
+/// A Stream of batchs.
+pub struct TimedRequestBatcherState<St: Stream> {
+    stream: Pin<Box<Fuse<St>>>,
+    cap: usize,
+    duration: Duration,
+    buffer_source: BufferSource,
+}
+
+impl<'a, St: Stream> TimedRequestBatcherState<St>
+where
+    St::Item: IngestLineSerialize<String, bytes::Bytes, HashMap<String, String>, Ok = ()>
+        + GetOffset
+        + std::marker::Send
+        + std::marker::Sync
+        + 'a,
+{
+    pub fn new(stream: St, capacity: usize, duration: Duration) -> TimedRequestBatcherState<St> {
+        assert!(capacity > 0);
+
+        // TODO expose parameters
+        let buffer_source = Box::pin(body_serializer_source(
+            16 * 1024, /* 16 KB segments */
+            50,        /* 16KB * 50 = 256 KB initial capacity */
+            None,      /* No max size */
+            Some(100), /* max 512KB idle buffers */
+        ));
+
+        TimedRequestBatcherState {
+            stream: Box::pin(stream.fuse()),
+            cap: capacity,
+            duration,
+            buffer_source,
+        }
+    }
+}
+
+pub fn timed_request_batch_stream<'a>(
+    stream: impl Stream<
+            Item = impl IngestLineSerialize<String, bytes::Bytes, HashMap<String, String>, Ok = ()>
+                       + GetOffset
+                       + std::marker::Send
+                       + std::marker::Sync
+                       + 'a,
+        > + 'a,
+    capacity: usize,
+    duration: Duration,
+) -> TimedRequestBatchStream<'a> {
+    TimedRequestBatchStream::new(Box::pin(stream::unfold(
+        TimedRequestBatcherState::new(stream, capacity, duration),
+        |mut state| async move {
+            // Get a new buffer
+            let buffer = state.buffer_source.next().await;
+            let mut current: IngestBodySerializer = match buffer {
+                Some(buffer) => match buffer {
+                    Ok(buffer) => buffer,
+                    Err(e) => return Some((Err(e.into()), state)),
+                },
+                None => return Some((Err(TimedRequestBatcherError::BufferStreamError), state)),
+            };
+
+            let mut offsets: OffsetMap = OffsetMap::default();
+
+            // Set the initial timeout
+            let mut timeout = Delay::new(state.duration);
+
+            loop {
+                let item = select(state.stream.next(), &mut timeout).await;
+
+                match item {
```

Comment:
NIT: Isn't `tokio::select!()` macro better suited for this?

for example:

```rust
loop {
    let mut delay = time::delay_for(DURATION_SAMPLE);

    loop {
        tokio::select! { /*delay or task*/}
    }
}
```

Reply:
We can yeh, tokio is already a dependency in the http module so we don't need to stick to futures here